README
Dezi

Dezi is a search platform based on Apache Lucy, Swish3, Plack and Search::OpenSearch.
Dezi integrates several search toolkits into an easy-to-use server.

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Dezi

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Dezi

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Dezi

    CPAN Ratings
        http://cpanratings.perl.org/d/Dezi

    Search CPAN
        http://search.cpan.org/dist/Dezi/


COPYRIGHT AND LICENCE

Copyright (C) 2011 Peter Karman

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Dezi

0.004003 15 May 2018
 - fix pod bugs in Dezi::Config

0.004002 30 April 2015
 - fix bin/dezi --auto-commit logic to default to undef so that the
   Engine config can be set via config file.

0.004001 1 Sept 2014
 - allow for fully-qualified URLs in search_path, index_path, etc.
 - require SWISH::3 to (hopefully) solve some failing Moose constraints
   around locale-specific Num values (as with floats like search_time)
 - bump dependencies reflecting Lucy 0.4.0 release

0.004000 30 July 2014
 - switch all references to SWISH::Prog to Dezi::App
 - require Dezi::App 0.006 or later instead of SWISH::Prog

0.003000 07 June 2014
 - official Moose release in support of S::OS 0.400

0.002998_01 05 June 2014
 - call to_app explicitly on Search::OpenSearch::Server instances
 - use Moose (as all Search:: dependencies use it now)

0.002012 27 Feb 2014
 - fix Config docs re: Dezi::Admin, add auto_commit note
 - add --auto_commit (--ac) option to toggle auto_commit via cli

0.002011 13 Nov 2013
 - fix bug in Server::About where fields is not populated

0.002010 12 Feb 2013
 - change --elastic to use 'autoall' with SWISH::3 1.000006

0.002009 12 Feb 2013
 - add --elastic feature to dezi cli
 - document strip_markup => 1 in snipper_config

0.002008 8 Feb 2013
 - fix about->admin_class reporting
 - add --admin option to dezi cli, document --admin-class

0.002007 5 Feb 2013
 - fix internal config passing

0.002006 3 Feb 2013
 - internal changes to support Dezi::Admin
 - internal changes to support Dezi::MultiTenant

0.002005 18 Dec 2012
 - fix pod SYNOPSIS for About
 - add require_root param to About->new

0.002004 17 Oct 2012
 - add LucyX::Suggester support for SOS 0.23 and SOS::Engine::Lucy 0.13

0.002003 15 Oct 2012
 - add Tiny response format support per SOS 0.21.
 - add --ui shortcut to dezi cli.

0.002002 13 Sept 2012
 - fix security issue with non-idempotent requests to /search URL app,
   bypassing authentication on the /index URL app. Dezi::Config now has a
   search_server instance and an index_server instance instead of a single
   server instance. Requires Search::OpenSearch::Server 0.19.

0.002001 12 Sept 2012
 - documentation fixes

0.002000 11 Sept 2012 
 - refactor all config into Dezi::Config
 - add optional basic authentication for /index, /commit and /rollback
   actions
 - add Architecture doc, flesh out Tutorial doc.

0.001008 3 Sept 2012
 - add /commit and /rollback features per Search::OpenSearch::Server 0.17

0.001007 21 Aug 2012
 - pod additions and fixes.
 - quote some objects in About to force stringify.

0.001006 17 Aug 2012
 - support SPORE specification in About
   (https://github.com/SPORE/specifications/blob/master/spore_description.p
   od)
 - fixes to URI generation in About
 - now requires Plack::Middleware::ReverseProxy for automatic detection
   when running behind a reverse proxy.

0.001005 10 July 2012
 - add config documentation
 - refactor Dezi::Server to create Dezi::Server::About class

0.001004 16 March 2012
 - added ui-class and admin-class options

0.001003 22 Oct 2011
 - fix dependencies in Makefile.PL
 - add Dezi::Tutorial

0.001002 29 Sept 2011
 - fix docs, default engine config
 - add / about() feature in support of Dezi::Client

0.001001 02 Aug  2011
 - change url root to /search, reserving namespaces /index and /admin as
   well

0.001000 21 June 2011
 - Initial release.



t/example-config.pl
# example Dezi server config file, in Perl format
# Dezi uses Config::Any, so any format supported by
# that module should work as well.

use CHI;

my $c = {

    engine_config => {

        # name of the index(es)
        index => [qw( t/test.index )],

        # which facets to calculate, and how many results to consider
        facets => {
            names       => [qw( color size flavor )],
            sample_size => 10_000,
        },

        # result attributes in response
        fields => [qw( color size flavor )],

        # options passed to SWISH::Prog::Lucy::Indexer->new
        indexer_config => {
            config => { 

                # searchable fields
                MetaNames => 'color size flavor',

                # attributes to store
                PropertyNames => 'color size flavor',

                # auto-vivify new fields based on POSTed docs
                UndefinedMetaTags => 'auto',

                # treat unknown mime types as text/plain
                DefaultContents => 'TXT',

                # use English snowball stemmer
                FuzzyIndexingMode => 'Stemming_en1',

            }, 

            # store token positions to optimize snippet creation
            highlightable_fields => 1,
        },

        # options passed to SWISH::Prog::Lucy::Searcher->new
        searcher_config => {
            max_hits             => 1000,
            find_relevant_fields => 1,
        },

        # cache facets for speed-up
        cache => CHI->new(
            driver           => 'File',
            dir_create_mode  => 0770,
            file_create_mode => 0660,
            root_dir         => "/tmp/opensearch_cache",
        ),
        cache_ttl => 3600,

        # explicitly turn off highlighting for some fields
        do_not_hilite => [qw( color )],

        # see Search::Tools::Snipper
        snipper_config => { as_sentences => 1 },

        # see Search::Tools::HiLiter
        hiliter_config => { class => 'h', tag => 'b' },

        # see Search::Query::Parser
        parser_config => {},

    }
};


Makefile.PL
use strict;
use warnings;
use inc::Module::Install;
use 5.008003;

name 'Dezi';
all_from 'lib/Dezi.pm';
requires 'Test::More'                       => 0;
requires 'Search::Tools'                    => '1.000';
requires 'Search::Query'                    => '0.300';
requires 'Search::Query::Dialect::Lucy'     => '0.201';
requires 'Lucy'                             => 0.003002;
requires 'Dezi::App'                        => 0.013;
requires 'SWISH::3'                         => '1.000013';
requires 'Search::OpenSearch::Engine::Lucy' => '0.400';
requires 'Search::OpenSearch::Server'       => '0.300';
requires 'Search::OpenSearch'               => '0.400';
requires 'Moose'                            => 0;
requires 'Type::Tiny'                       => 0;
requires 'Plack'                            => 0;
requires 'Class::Load'                      => 0;
requires 'Plack::Middleware::ReverseProxy'  => 0;
requires 'Plack::Middleware::Auth::Basic'   => 0;

# can't require in case Text::Aspell not available
recommends 'LucyX::Suggester' => '0.003';

perl_version '5.8.3';
license 'Perl_5';
homepage 'https://github.com/karpet/Dezi';
bugtracker 'http://rt.cpan.org/NoAuth/Bugs.html?Dist=Dezi';
repository 'http://github.com/karpet/Dezi';
install_script 'bin/dezi';

WriteMakefile(
    PL_FILES => {},
    dist     => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 't/test.index Dezi-* dezi.index' },
);


lib/Dezi.pm
package Dezi;
use warnings;
use strict;

our $VERSION = '0.004003';

1;

__END__

=head1 NAME

Dezi - REST search platform

=head1 SYNOPSIS

Start the Dezi server, listening on port 5000:

 % dezi -p 5000

Add a document B<foo> to the index:

 % curl http://localhost:5000/index/foo -XPOST \
   -d '<doc><title>bar</title>hello world</doc>' \
   -H 'Content-Type: application/xml'
   
Search the index for B<bar>:

 % curl 'http://localhost:5000/search?q=bar&t=JSON'
 % curl 'http://localhost:5000/search?q=bar&t=XML'
 
Read the usage statement:

 % dezi -h

=head1 DESCRIPTION

Dezi is a search platform based on Apache L<Lucy>, Swish3,
L<Search::OpenSearch> and L<Search::Query>. 

Dezi integrates several CPAN search libraries into one
easy-to-use interface.

This document is a placeholder for the namespace and documentation
only. You should read:

=over

=item

the L<Dezi::Tutorial>

=item 

the L<Dezi::Config> documentation

=item

the L<Dezi::Server> class documentation

=item

the L<Dezi::Architecture>

=item

the L<Search::OpenSearch::Server::Plack> documentation, upon which
Dezi relies heavily.

=back

=head1 AUTHOR

Peter Karman, C<< <karman at cpan.org> >>

=head1 ACKNOWLEDGEMENTS

Many of the code ideas in Dezi were developed
while working on search-related projects for the
Public Insight Network at American Public Media.
L<http://www.publicinsightnetwork.org>.

=head1 BUGS

Please report any bugs or feature requests to C<bug-dezi at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Dezi>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Dezi


You can also look for information at:

=over 4

=item * Website

L<http://dezi.org/>

=item * IRC

#dezisearch at freenode

=item * Mailing list

L<https://groups.google.com/forum/#!forum/dezi-search>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Dezi>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Dezi>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Dezi>

=item * Search CPAN

L<http://search.cpan.org/dist/Dezi/>

=back

=head1 COPYRIGHT & LICENSE

Copyright 2011-2018 Peter Karman.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=head1 SEE ALSO

L<Search::OpenSearch>, L<SWISH::3>, L<SWISH::Prog::Lucy>,
L<Plack>, L<Lucy>, L<http://dezi.org>

=cut


lib/Dezi/Server.pm
package Dezi::Server;
use Moose;
extends 'Search::OpenSearch::Server::Plack';
use Carp;
use Plack::Builder;
use Dezi::Server::About;
use Dezi::Config;
use Scalar::Util qw( blessed );

our $VERSION = '0.004003';

sub app {
    my ( $class, $config ) = @_;

    my $dezi_config;
    if ( blessed $config) {
        $dezi_config = $config;
    }
    else {
        $dezi_config
            = Dezi::Config->new( { %$config, server_class => $class } );
    }

    return builder {

        enable "SimpleLogger",
            level => $dezi_config->debug ? "debug" : "warn";

        enable_if { $_[0]->{REMOTE_ADDR} eq '127.0.0.1' }
        "Plack::Middleware::ReverseProxy";

        mount $dezi_config->search_path() =>
            $dezi_config->search_server->to_app;
        mount $dezi_config->index_path() => builder {
            if ( defined $dezi_config->authenticator ) {
                enable "Auth::Basic",
                    authenticator => $dezi_config->authenticator,
                    realm         => 'Dezi Indexer';
            }
            $dezi_config->index_server->to_app;
        };
        mount $dezi_config->commit_path() => builder {
            if ( defined $dezi_config->authenticator ) {
                enable "Auth::Basic",
                    authenticator => $dezi_config->authenticator,
                    realm         => 'Dezi Indexer';
            }
            sub {
                my $env = shift;
                if ( $env->{REQUEST_METHOD} eq 'POST' ) {
                    $env->{REQUEST_METHOD} = 'COMMIT';
                }
                $dezi_config->index_server->call($env);
            };
        };
        mount $dezi_config->rollback_path() => builder {
            if ( defined $dezi_config->authenticator ) {
                enable "Auth::Basic",
                    authenticator => $dezi_config->authenticator,
                    realm         => 'Dezi Indexer';
            }
            sub {
                my $env = shift;
                if ( $env->{REQUEST_METHOD} eq 'POST' ) {
                    $env->{REQUEST_METHOD} = 'ROLLBACK';
                }
                $dezi_config->index_server->call($env);
            };
        };

        if ( $dezi_config->ui ) {
            mount $dezi_config->ui_path() => $dezi_config->ui->to_app;
        }

        if ( $dezi_config->admin ) {
            mount $dezi_config->admin_path() => $dezi_config->admin;
        }

        # default is just self-description
        mount '/' => sub {
            my $req = Plack::Request->new(shift);
            return Dezi::Server::About->new(
                require_root  => 1,
                server        => $dezi_config->index_server,
                request       => $req,
                search_path   => $dezi_config->search_path,
                index_path    => $dezi_config->index_path,
                commit_path   => $dezi_config->commit_path,
                rollback_path => $dezi_config->rollback_path,
                admin_path    => $dezi_config->admin_path,
                ui_path       => $dezi_config->ui_path,
                config        => $dezi_config,
                version       => $VERSION,
                base_uri      => $dezi_config->base_uri,
            );
        };

        mount '/favicon.ico' => sub {
            my $req = Plack::Request->new(shift);
            my $res = $req->new_response();
            $res->redirect( 'http://dezi.org/favicon.ico', 301 );
            $res->finalize();
        };

    };

}

1;

__END__

=head1 NAME

Dezi::Server - Dezi Plack server

=head1 SYNOPSIS

Start the Dezi server, listening on port 5000:

 % dezi -p 5000

Add a document B<foo> to the index:

 % curl http://localhost:5000/index/foo -XPOST \
   -d '<doc><title>bar</title>hello world</doc>' \
   -H 'Content-Type: application/xml'
   
Search the index:

 % curl 'http://localhost:5000/search?q=bar&t=JSON'
 % curl 'http://localhost:5000/search?q=bar&t=XML'

=head1 DESCRIPTION

Dezi is a search platform based on Apache Lucy, Swish3,
Search::OpenSearch and Search::Query. 

Dezi integrates several CPAN search libraries into one
easy-to-use interface.

Be sure to read the perldoc for L<Search::OpenSearch::Engine>
and L<Search::OpenSearch::Server::Plack>.

=head1 METHODS

Dezi::Server is a subclass of L<Search::OpenSearch::Server::Plack>.
Only new methods are overridden.

=head2 app( I<config> )

Class method that uses L<Plack::Builder> to construct the server
application. I<config> should be a hashref that is converted
internally to a L<Dezi::Config> object.

Read the L<Dezi::Config> documentation for full reference of all
options available in I<config>.

Returns the Plack $app via the L<Plack::Builder> builder() function.

=head1 AUTHOR

Peter Karman, C<< <karman at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-dezi at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Dezi>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Dezi::Server


You can also look for information at:

=over 4

=item * Website

L<http://dezi.org/>

=item * IRC

#dezisearch at freenode

=item * Mailing list

L<https://groups.google.com/forum/#!forum/dezi-search>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Dezi>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Dezi>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Dezi>

=item * Search CPAN

L<http://search.cpan.org/dist/Dezi/>

=back

=head1 COPYRIGHT & LICENSE

Copyright 2011-2018 Peter Karman.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=head1 SEE ALSO

L<Search::OpenSearch>, L<SWISH::3>, L<Dezi::App>,
L<Plack>, L<Lucy>

=cut


lib/Dezi/Config.pm
package Dezi::Config;
use Moose;
use Types::Standard qw( InstanceOf Str Bool CodeRef Maybe HashRef );
use Carp;
use Data::Dump qw( dump );
use Class::Load;

has 'search_path' => ( is => 'rw', isa => Str, default => sub {'/search'} );
has 'index_path'  => ( is => 'rw', isa => Str, default => sub {'/index'} );
has 'commit_path' => ( is => 'rw', isa => Str, default => sub {'/commit'} );
has 'rollback_path' =>
    ( is => 'rw', isa => Str, default => sub {'/rollback'} );
has 'ui_path'    => ( is => 'rw', isa => Str, default => sub {'/ui'} );
has 'admin_path' => ( is => 'rw', isa => Str, default => sub {'/admin'} );
has 'ui'         => (
    is      => 'rw',
    isa     => Maybe [ InstanceOf ['Dezi::UI'] ],
    lazy    => 1,
    builder => 'init_ui',
);
has 'admin' => (
    is      => 'rw',
    isa     => Maybe [CodeRef],
    lazy    => 1,
    builder => 'init_admin',
);
has 'server_class' =>
    ( is => 'rw', isa => Str, default => sub {'Dezi::Server'} );
has 'ui_class'    => ( is => 'rw', isa => Maybe [Str] );
has 'admin_class' => ( is => 'rw', isa => Maybe [Str] );
has 'debug' =>
    ( is => 'rw', isa => Bool, default => sub { $ENV{DEZI_DEBUG} || 0 } );
has 'base_uri' => ( is => 'rw', isa => Str, default => sub {''} );
has 'search_server' => (
    is      => 'rw',
    isa     => InstanceOf ['Search::OpenSearch::Server::Plack'],
    lazy    => 1,
    builder => 'init_search_server',
);
has 'index_server' => (
    is      => 'rw',
    isa     => InstanceOf ['Search::OpenSearch::Server::Plack'],
    lazy    => 1,
    builder => 'init_index_server',
);
has 'authenticator' => ( is => 'rw', isa => Maybe [CodeRef] );
has 'server_config' => ( is => 'rw', isa => HashRef );

our $VERSION = '0.004003';

sub init_ui {
    my $self = shift;
    if ( $self->ui_class ) {
        Class::Load::load_class $self->ui_class;
        return $self->ui_class->new(
            search_path => $self->search_path,
            base_uri    => $self->base_uri
        );
    }
    return undef;
}

sub init_admin {
    my $self = shift;
    if ( $self->admin_class ) {
        Class::Load::load_class $self->admin_class;
        return $self->admin_class->app(
            user_config => $self->server_config,
            searcher    => $self->search_server,
            base_uri    => $self->base_uri,
        );
    }
    return undef;
}

sub BUILDARGS {
    my $class = shift;
    my %args;
    if ( @_ == 1 and ref( $_[0] ) eq 'HASH' ) {
        %args = %{ $_[0] };
    }
    else {
        %args = @_;
    }

    # save credentials in a closure but not plain in args
    my $username      = delete $args{username};
    my $password      = delete $args{password};
    my $authenticator = ( defined $username and defined $password )
        ? sub {
        my ( $u, $p ) = @_;
        return $u eq $username && $p eq $password;
        }
        : undef;
    $args{authenticator} ||= $authenticator;

    # save anything we do not have an explicit method for
    # in the server_config stash
    $args{server_config} ||= {};
    for my $arg ( keys %args ) {
        if ( !$class->can($arg) ) {
            $args{server_config}->{$arg} = delete $args{$arg};
        }
        elsif ( $arg eq 'admin' and ref( $args{$arg} ) eq 'HASH' ) {

            # special case. compatability with Dezi::Admin::Config.
            $args{server_config}->{$arg} = delete $args{$arg};
        }
    }

    # make sure all paths are /-prefixed
    for my $p (
        qw( search_path index_path commit_path rollback_path ui_path admin_path )
        )
    {
        if ( exists $args{$p} ) {
            $args{$p} = "/$args{$p}" unless $args{$p} =~ m!^(/|https?:)!;
        }
    }
    return \%args;
}

sub BUILD {
    my $self          = shift;
    my $server_class  = $self->server_class;
    my $search_path   = $self->search_path;
    my $index_path    = $self->index_path;
    my $commit_path   = $self->commit_path;
    my $rollback_path = $self->rollback_path;
    my $ui_path       = $self->ui_path;
    my $admin_path    = $self->admin_path;
    my $base_uri      = $self->base_uri;

    Class::Load::load_class $server_class;
    my $search_server = $server_class->new(
        %{ $self->server_config },
        engine_config => $self->apply_default_engine_config(
            { %{ $self->server_config }, search_path => $search_path }
        ),
        http_allow => [qw( GET )],
    );
    my $index_server = $server_class->new(
        %{ $self->server_config },
        engine_config => $self->apply_default_engine_config(
            { %{ $self->server_config }, search_path => $search_path }
        ),
    );

    $self->search_server($search_server);
    $self->index_server($index_server);

    $self->debug and carp dump $self;

    return $self;
}

sub apply_default_engine_config {
    my ( $self, $args ) = @_;
    my $engine_config = $args->{engine_config} || {};
    $engine_config->{type}  ||= 'Lucy';
    $engine_config->{index} ||= ['dezi.index'];
    my $search_path = delete $args->{search_path};
    $engine_config->{link} ||= $search_path;
    $engine_config->{default_response_format} ||= 'JSON';
    $engine_config->{debug} = $args->{debug} || $self->debug;
    return $engine_config;
}

sub as_hash {
    my $self = shift;
    return %$self;
}

1;

__END__

=pod

=head1 NAME

Dezi::Config - Dezi server configuration

=head1 SYNOPSIS

 use Dezi::Config;
 use CHI;         # optional, see cache below
 use Dezi::UI;    # optional, see ui_class and ui below
 use Dezi::Admin; # optional, see admin_class and admin below
 use Dezi::Stats; # optional, see stats_logger below

 my $dezi_config = Dezi::Config->new({
 
    search_path     => '/search',
    index_path      => '/index',
    commit_path     => '/commit',
    rollback_path   => '/rollback',
    ui_path         => '/ui',
    ui_class        => 'Dezi::UI',
    # or
    # ui              => Dezi::UI->new()
   
    admin_path      => '/admin', 
    admin_class     => 'Dezi::Admin',
    # or
    # admin           => Dezi::Admin->new(),
    
    base_uri        => '',
    server_class    => 'Dezi::Server',
    
    # authentication for non-idempotent requests.
    # if both username && password are defined,
    # then /index, /commit and /rollback require
    # basic authentication credentials.
    username        => 'someone',
    password        => 'somesecret',
    
    # optional
    # see Dezi::Stats
    stats_logger => Dezi::Stats->new(
        type        => 'DBI',
        dsn         => 'DBI::mysql:database=mydb;host=localhost;port=3306',
        username    => 'myuser',
        password    => 'mysecret',
    ),
    
    # see Search::OpenSearch::Engine
    engine_config => {

        default_response_format => 'JSON',
        
        # could be any Search::OpenSearch::Engine::* class
        type    => 'Lucy',

        # name of the index(es)
        index   => [qw( path/to/your.index )],

        # which facets to calculate, and how many results to consider
        facets => {
            names       => [qw( color size flavor )],
            sample_size => 10_000,
        },

        # result attributes in response
        fields => [qw( color size flavor )],

        # options passed to indexer defined by Engine type (above)
        # defaults to Dezi::Lucy::Indexer->new
        indexer_config => {
        
            # see Dezi::Indexer::Config
            # and https://dezi.org/swish-e-docs/SWISH-CONFIG.pod.html
            config => { 

                # searchable fields
                MetaNames => 'color size flavor',

                # attributes to store
                PropertyNames => 'color size flavor',

                # auto-vivify new fields based on POSTed docs.
                # use this if you want ElasticSearch-like effect.
                UndefinedMetaTags => 'auto',

                # treat unknown mime types as text/plain
                DefaultContents => 'TXT',

                # use English snowball stemmer
                FuzzyIndexingMode => 'Stemming_en1',

            }, 

            # store token positions to optimize snippet creation
            highlightable_fields => 1,
        },

        # options passed to searcher defined by Engine type (above)
        # defaults to Dezi::Lucy::Searcher->new
        searcher_config => {
            max_hits             => 1000,
            find_relevant_fields => 1,
            qp_config => {
                dialect   => 'Lucy',
                null_term => 'NULL',
                # see Search::Query::Parser and Search::Query::Dialect::Lucy
                # for full list of options
            },
        },

        # see LucyX::Suggester
        suggester_config => {
            limit  => 10,
            fields => [qw( color size )],

            # passed to Search::Tools::Spellcheck->new
            # along with parser_config
            spellcheck_config => {
                lang => 'en_US',
            },
        },

        # cache facets for speed-up.
        # this is the Search::OpenSearch default setting
        cache => CHI->new(
            driver           => 'File',
            dir_create_mode  => 0770,
            file_create_mode => 0660,
            root_dir         => "/tmp/opensearch_cache",
        ),
        
        # how long should the facet cache live
        # each cache entry is per-unique-query
        cache_ttl => 3600,

        # explicitly turn off highlighting for some fields
        do_not_hilite => { color => 1 },

        # see Search::Tools::Snipper
        snipper_config => { as_sentences => 1, strip_markup => 1, },

        # see Search::Tools::HiLiter
        hiliter_config => { class => 'h', tag => 'b' },

        # see Search::Tools::QueryParser
        parser_config => {},

        # see Search::OpenSearch::Engine::Lucy
        auto_commit => 1, # set to 0 to enable transactions with /commit and /rollback

    }
 
 });
 
=head1 DESCRIPTION

Dezi::Config parses configuration settings, applies default values,
and instantiates component objects for L<Dezi::Server>. 

Mostly this class exists in order to document, in one location, 
all the options available for the Dezi server. 
You will rarely use Dezi::Config directly; it is intended
as an internal class for use by Dezi::Server. Instead, your C<dezi_config.pl> file
contents are parsed by Dezi::Config and applied to the server application.

The SYNOPSIS section provides all the default configuration values,
with comments indicating where more complete documentation may
be available for the relevant components. The rest of the documentation
below is specific to this class and probably B<not> what you're looking for
as a Dezi user.

=head1 METHODS

=head2 new( I<hashref> )

See the SYNOPSIS for a complete description of the keys/values supported
in I<hashref>.

The following I<hashref> keys are also supported as accessor/mutator
methods on the object returned from new():

    search_path
    index_path
    commit_path
    rollback_path
    ui_path
    admin_path
    ui
    admin
    debug
    base_uri
    search_server
    index_server

=head2 BUILD

Internal method called by new().

=head2 BUILDARGS

Internal method. Some convenient arg munging for new().

=head2 init_ui

Returns an instance of B<ui_class> if set.

=head2 init_admin

Returns an instance of B<admin_class> if set.

=head2 apply_default_engine_config( I<hashref> )

Default L<Search::OpenSearch::Engine> options are applied directly to I<hashref>.
This method is called internally by new().

=head2 authenticator

If I<username> and I<password> are passed to new(),
the authenticator() method will return a CODE ref for passing
to L<Plack::Middleware::Auth::Basic>.

=head2 as_hash

Returns the object as a plain Perl hash of key/value pairs.

=head1 AUTHOR

Peter Karman, C<< <karman at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-dezi at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Dezi>.  
I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Dezi::Config


You can also look for information at:

=over 4

=item * Website

L<http://dezi.org/>

=item * IRC

#dezisearch at freenode

=item * Mailing list

L<https://groups.google.com/forum/#!forum/dezi-search>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Dezi>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Dezi>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Dezi>

=item * Search CPAN

L<http://search.cpan.org/dist/Dezi/>

=back

=head1 COPYRIGHT & LICENSE

Copyright 2012-2018 Peter Karman.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=head1 SEE ALSO

L<Search::OpenSearch>, L<Search::Tools>, L<SWISH::3>, L<Dezi::App>,
L<Plack>, L<Lucy>

=cut
